test/extensions/filters/http/transformation/transformation_filter_test.cc

```diff
@@ -43,34 +50,122 @@ class TransformationFilterTest : public testing::Test {
+    initFilterWithBodyTemplate(configType, "solo");
 
-    auto &transformation = (*route_config_.mutable_request_transformation());
-    transformation.mutable_header_body_transform();
+    EXPECT_CALL(filter_callbacks_, addDecodedData(_, false)).Times(1);
+    EXPECT_CALL(filter_callbacks_, clearRouteCache()).Times(0);
+    auto res = filter_->decodeHeaders(headers_, true);
+    EXPECT_EQ(Http::FilterHeadersStatus::Continue, res);
+    EXPECT_EQ(val, config_->stats().request_header_transformations_.value());
+  }
+
+  void transformsOnHeadersAndClearCache(TransformationFilterTest::ConfigType configType, unsigned int val) {
+    initFilterWithBodyTemplate(configType, "solo");
+    route_config_.set_clear_route_cache(true);
     initFilter(); // Re-load config.
+
+    EXPECT_CALL(filter_callbacks_, addDecodedData(_, false)).Times(1);
+    EXPECT_CALL(filter_callbacks_, clearRouteCache()).Times(1);
+    auto res = filter_->decodeHeaders(headers_, true);
+    EXPECT_EQ(Http::FilterHeadersStatus::Continue, res);
+    EXPECT_EQ(val, config_->stats().request_header_transformations_.value());
   }
 
+  void transformsResponseOnHeaders(unsigned int val) {
+    route_config_.mutable_response_transformation()
+      ->mutable_transformation_template()
+      ->mutable_body()
+      ->set_text("solo");
+
+    initFilter();
+
+    filter_->decodeHeaders(headers_, true);
+    EXPECT_CALL(encoder_filter_callbacks_, addEncodedData(_, false)).Times(1);
+    auto res = filter_->encodeHeaders(headers_, true);
+    EXPECT_EQ(Http::FilterHeadersStatus::Continue, res);
+    EXPECT_EQ(val , config_->stats().response_header_transformations_.value());
+  }
+
+  void happyPathWithBody(TransformationFilterTest::ConfigType configType, unsigned int val) {
+      initFilterWithBodyTemplate(configType, "{{a}}");
+
+    auto resheaders = filter_->decodeHeaders(headers_, false);
+    ASSERT_EQ(Http::FilterHeadersStatus::StopIteration, resheaders);
+
+    std::string upstream_body;
+    EXPECT_CALL(filter_callbacks_, addDecodedData(_, false))
+        .WillOnce(Invoke(
+            [&](Buffer::Instance &b, bool) { upstream_body = b.toString(); }));
+
+    Buffer::OwnedImpl downstream_body("{\"a\":\"b\"}");
+    auto res = filter_->decodeData(downstream_body, true);
+    EXPECT_EQ(Http::FilterDataStatus::Continue, res);
+    EXPECT_EQ("b", upstream_body);
+    EXPECT_EQ(val, config_->stats().request_body_transformations_.value());
+  }
+  
+  void happyPathWithBodyPassthrough(TransformationFilterTest::ConfigType configType) {
+    initFilterWithBodyPassthrough(configType);
+
+    auto resheaders = filter_->decodeHeaders(headers_, false);
+    EXPECT_EQ(Http::FilterHeadersStatus::Continue, resheaders);
+
+    Buffer::OwnedImpl downstream_body("{\"a\":\"b\"}");
+    auto res = filter_->decodeData(downstream_body, true);
+    EXPECT_EQ(Http::FilterDataStatus::Continue, res);
+    EXPECT_EQ(0U, config_->stats().request_body_transformations_.value());
```

Comment:
should the header stats be not zero here? can we test that?

Reply:
Sure